### DISCUSSION TITLE
Would a tiny ARMI DB be a useful addition for testing?

### DISCUSSION THREAD

**Discussion by opotowsky** [terrapower/armi#1385] [Ideas]
In [this PR](https://github.com/terrapower/armi/pull/1371#issue-1833885314), @john-science brought up the fact that a small DB could improve testing for CLI endpoints. 

But I think having a DB might be beneficial for simplifying/improving many tests, or even adding new ones.

Is this a worthwhile effort? How small can we make this and have it still be useful? Feelings aren't hurt if this is shot down, just didn't want to lose the thought in a merged PR.

@keckler @albeanth @mgjarrett @onufer

**Comment by albeanth** [comment]
Not sure if this sort of check already exists, but I think it would be useful to include a small database to ensure backwards compatibility with the most recent tag. E.g., on CI for each PR we could run
```python
db = databaseFactory(pathToH5, "r")
with db:
    r = db.load(0, 0, allowMissing=True)
```

**Comment by john-science** [comment]
I realized that we have several tests that already have an ARMI DB for testing, they build it on the fly from the test reactor:

https://github.com/terrapower/armi/blob/2dab1222f7ea79d3083db3aabf35b00510bf8d2a/armi/bookkeeping/db/tests/test_database3.py#L255-L267

I don't really _love_ the idea of having a static binary file in the repo that can just go stale and bad at any time.

I think whoever wrote all these DB tests had the right idea.

**Reply by albeanth** [reply]
@john-science could these tests be configured to test backwards compatibility? Given that downstream analyses can include retrieving information off of a database from a previous tagged release, I think it's important to be able to ensure backwards compatibility. 

I think the idea of a static binary file going "stale" or "bad" is the whole idea behind the test. If it goes "bad" then we've broken backwards compatibility and either need to 1) provide a conversion/migration script for affected versions, or 2) stop supporting those releases. 

What do you think?

**Reply by jakehader** [reply]
I think it might be interesting to consider serializing /  deserializing the database into a text based format for the purposes of testing and doing comparisons of differences. 

This would mean that the database interface and database 3, would need to expand to support say a .json format but in theory in could be done. I think the critical issue with database changes within the API is related to parameter changes (adding new ones that don't exist in a previous version or changing names). There may be other hdf5 dependency impacts over time or h5py changes, but those should be infrequent.

If it's the parameter state that is changing between two versions it might be easier to create an application based parameter test that on the first time run would create a catalog of the parameter definition metadata and store that into a file (if enabled/configured). If the file already exists then a test would regenerate the state information for the new code version, compare to an existing version and report pass/fail. If it's the first time an application runs this then it would just create the file and the application owner could decide to track it or not. Maybe not an elegant solution, but maybe lighter weight than storage a full database and trying to put parameter data on it that could become stale.

**Reply by john-science** [reply]
# Various Thoughts

## 1. Old Migration Scripts

Backwards compatibility in ARMI right now is pretty hindered.  We already have some backwards migrations scripts for all versions of the code:

https://github.com/terrapower/armi/tree/main/armi/scripts/migration

But I don't have any example DBs to test those on.  Not even OUTSIDE the repo.

## 2. New Migration Scripts

We have some new migration scripts in-house I considered brining into ARMI, but they are not all general enough to work for a wider audience.

## 3. Including Migration Scripts in ARMI

I'm happy to have the ARMI migration scripts in the ARMI repo. But the fact that they are INSTALLED and importable as part of ARMI is strange and awkward.  We will never have code coverage on them, so why bother?

I would prefer to have them outside the installable `armi/` package/directory.

I have even considered making them their own repo, so we can include nice, full DBs with them.

**Comment by onufer** [comment]
Late to the party. I like the idea. the main reason people don't like this sort of idea is database could be big and make git more bulky. i trust folks to make it small enough.

even if its only 3 assemblies its at least something to do more tests on. it also might help us see backwards compatibly problems sooner.
